### PR TITLE
Add Certificate Authentication Methods

### PR DESCRIPTION
**Pull request by Tylerlhess**
Added Certificate Authentication

### CODE REVIEW

**Comment by jeffwecan** [comment]
Thanks for the contribution @Tylerlhess! Do you have any interest in attempting to include some [documentation](https://github.com/hvac/hvac/blob/develop/CONTRIBUTING.md#documentation) and/or [test coverage](https://github.com/hvac/hvac/blob/develop/CONTRIBUTING.md#testing) to cover this newly added TLS certificates auth method class? 😃 

(If not, that is perfectly okay, it may just take me a bit to see about filling into those aspects before I'd feel ready to merge this in.)

**Comment by Tylerlhess** [comment]
Oh I figured I needed to write the test cases myself. I just need to get the adapter to give me the cert and it should be passing

**Comment by Tylerlhess** [comment]
Helps to actually write tests and test code before submitting the PR. :)

File: hvac/v1/__init__.py
Comment (jeffwecan) [inline hvac/v1/__init__.py:595]: Since we're updating this extant call site to use the new class, could you please tack on a related decorator for this `auth_tls()` method? E.g.:

```python
    @utils.deprecated_method(
        to_be_removed_in_version='0.13.0',
        new_method=api.auth_methods.Cert.login,
    )
```

(This `Client` class is a bit unwieldy, so we've been taking any opportunity to move folks off of using its methods where possible. Related reference: https://github.com/hvac/hvac/blob/develop/CONTRIBUTING.md#backwards-compatibility-breaking-changes )

**Review by jeffwecan — APPROVED** [review]
Thanks so much for working to fill in some test cases here! Looking things over, there is just one bit of cleanup I'd be in favor of seeing before we merge -> release this.

**Comment by jeffwecan** [comment]
Ah shucks, looks like implementing that last suggestion may have exposed some gaps in our little `deprecated_method` handler 😂 :

```
_________ ERROR collecting tests/unit_tests/v1/test_aws_ec2_methods.py _________
tests/unit_tests/v1/test_aws_ec2_methods.py:6: in <module>
    from hvac import Client
hvac/__init__.py:1: in <module>
    from hvac.v1 import Client
hvac/v1/__init__.py:18: in <module>
    class Client(object):
hvac/v1/__init__.py:589: in Client
    def auth_tls(self, mount_point='cert', use_token=True):
hvac/utils.py:188: in decorator
    new_docstring=dedent(new_method.__doc__),
/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/textwrap.py:430: in dedent
    text = _whitespace_only_re.sub('', text)
E   TypeError: expected string or bytes-like object
```

My intuition is this is due to some logic related to setting `__doc__` on the decorated method combined with the `Cert.login()` method not having any docstrings? I'll scope that out to confirm and get back to ya!

**Comment by jeffwecan** [comment]
Alrighty, so I'll merge in https://github.com/hvac/hvac/pull/692 shortly and you could then rebase this branch on top of `develop` / merge develop into your branch to get those unrelated-to-your-changes test failures cleared.

**Comment by Tylerlhess** [comment]
The Issue was no doc string in the new method replacing the deprecated method. More of a bug in my code than in the decorator

**Comment by Tylerlhess** [comment]
Yeah hold off on merging this I might have to add setting the cert and ca directly into the sessions request. I also discovered that there may need to be a disclaimer that this will not work with the agent at present as it currently doesn't proxy the certificate request as far as I can see it would also need to be running in ssl. The agent can be used after certificate auth with the token. (This is what I had to do in my production environment as well as make the initial request through requests directly instead of being able to set the session variables needed to get the cert auth to work ) The current auth_tls method also has this failing currently and balks at passing the cacert kwarg down to the adapter even as the 'verify' kwarg. This is just the lack of passing kwargs down through to the adapter layer but I also want to make sure that my code is doing so properly. 
I am setting the kwargs to the adapter in my new_login method which should resolve the adapter getting these kwargs but I may also need to do this in the general login method. When I submitted the or I was under the impression that this was happening but it doesn't seem to be the case.

I don't think I will get the agent resolved in this pr as my experience with the agent is lacking

**Comment by jeffwecan** [comment]
I don't have much context on cert auth usage but I'm down to help scope out the agent angle as I have time over the next week or so.

**Comment by Tylerlhess** [comment]
OK this is now fully baked and ready to handle all the things.

File: hvac/api/auth_methods/cert.py
Comment (jeffwecan) [inline hvac/api/auth_methods/cert.py:91]: I think you'll want to wrap this in `utils.remove_nones()` to match similar methods' behavior (without that arrangement, it makes it harder for hvac users to update a single parameter without reverting the rest of them to their defaults generally):
```suggestion

        params = utils.remove_nones(
            {
                "name": name,
                "certificate": cert,
                "allowed_common_names": allowed_common_names,
                "allowed_dns_sans": allowed_dns_sans,
                "allowed_email_sans": allowed_email_sans,
                "allowed_uri_sans": allowed_uri_sans,
                "allowed_organizational_units": allowed_organizational_units,
                "required_extensions": required_extensions,
                "display_name": display_name,
                "token_ttl": token_ttl,
                "token_max_ttl": token_max_ttl,
                "token_policies": token_policies,
                "token_bound_cidrs": token_bound_cidrs,
                "token_explicit_max_ttl": token_explicit_max_ttl,
                "token_no_default_policy": token_no_default_policy,
                "token_num_uses": token_num_uses,
                "token_period": token_period,
                "token_type": token_type,
            }
        )
```

Also would need to add `from hvac import utils` to this file's imports to support that ^

**Review by jeffwecan — APPROVED** [review]
Looks good to me! Just had that one minor point / suggestion on the `create_ca_certificate_role()` method's params.

**Review by jeffwecan — APPROVED** [review]
Awesome, thanks again @Tylerlhess! 🎉

**Comment by jeffwecan** [comment]
FYI I'll aim to get a release including these additions shipped over the next day or so.

**Comment by HansAarneLiblik** [comment]
Since `hvac.Client` already accepts `cert` parameter, why is it needed to provide it to `auth.cert.login()` method again ? Can't it be retrieved from `hvac.Client` ?

File: hvac/api/auth_methods/cert.py
Comment (intgr) [inline hvac/api/auth_methods/cert.py:228]: Erm, this code is unreachable. `validate_pem_format()` is *documented* as returning `bool`, but it actually always returns `None`.

File: hvac/api/auth_methods/cert.py
Comment (intgr) [inline hvac/api/auth_methods/cert.py:235]: Iterating over a dict only yields keys, not key-value pairs. So this will always raise `ValueError` and the loop body is unreachable code.
```python
>>> for tls_part, value in {'cert_pem': '', 'key_pem': ''}:
...     pass

ValueError: too many values to unpack (expected 2)
```

File: hvac/api/auth_methods/cert.py
Comment (intgr) [inline hvac/api/auth_methods/cert.py:230]: `isinstance(e, type(exceptions.ParamValidationError()))`
->
`isinstance(e, exceptions.ParamValidationError)`
->
`except exceptions.ParamValidationError as e`

File: hvac/api/auth_methods/cert.py
Comment (intgr) [reply hvac/api/auth_methods/cert.py:228]: Note that the requests library does not accept a PEM string directly s argument -- it must always be a path. So this logic could never have worked anyway.

File: hvac/api/auth_methods/cert.py
Comment (intgr) [inline hvac/api/auth_methods/cert.py:210]: Fixed in https://github.com/hvac/hvac/pull/720

File: hvac/api/auth_methods/cert.py
Comment (intgr) [inline hvac/api/auth_methods/cert.py:253]: Shouldn't hvac exceptions derive from `exceptions.VaultError` class?

File: hvac/api/auth_methods/cert.py
Comment (xupefei) [reply hvac/api/auth_methods/cert.py:228]: The signature of `validate_pem_format` is `(param_name, param_argument)` but here it's being called like `(param_argument, param_name)` This validation will never pass :(

File: hvac/api/auth_methods/cert.py
Comment (Tylerlhess) [reply hvac/api/auth_methods/cert.py:210]: Good catch. The use_token was one of the things I hadn't originally had in scope as I wasn't originally planning on depricating the auth_tls method. I was originally only trying to create a method to create the certificate endpoints as well as allow an already created client to auth without the need to recreate it. This is because the certs are not saved as part of the adapter class and thus could only be used during instantiation of a new client object.

File: hvac/api/auth_methods/cert.py
Comment (Tylerlhess) [reply hvac/api/auth_methods/cert.py:253]: It does not let you subclass Exceptions in this way. It has similar patterns in other modules with module specific error messages.

File: hvac/api/auth_methods/cert.py
Comment (Tylerlhess) [reply hvac/api/auth_methods/cert.py:228]: Yes, this was from a miss reading of the requests verify params

:param verify: Either a boolean, in which case it controls whether we verify
            the server's TLS certificate, or a string, in which case it must be a path
            to a CA bundle to use

was read as 

:param verify: Either a boolean, in which case it controls whether we verify,
            the server's TLS certificate, or a string, in which case it must be a path
            to a CA bundle to use
            
I could never figure out why it thought that you could supply the tls cert as an argument but this logic was left in as it wasn't breaking. It wasn't working either.

as @xupefei pointed out I did also transpose the two arguments so this always failed.

File: hvac/api/auth_methods/cert.py
Comment (Tylerlhess) [reply hvac/api/auth_methods/cert.py:235]: fixed in #743

File: hvac/api/auth_methods/cert.py
Comment (intgr) [inline hvac/api/auth_methods/cert.py:69]: Reported bug #841 related to this code.

**Pull request merged**
